src/toxcore/dht/packet/errors.rs

```diff
@@ -0,0 +1,133 @@
+/*! Errors enum for DHT packets.
+*/
+
+use std::fmt;
+
+use failure::{Backtrace, Context, Fail};
+use nom::{Needed, ErrorKind as NomErrorKind};
+
+use std::convert::From;
+use std::io::Error as IoError;
+use std::io::ErrorKind as IoErrorKind;
+
+/// Error that can happen when calling `get_payload` of packet.
+#[derive(Debug)]
+pub struct GetPayloadError {
+    ctx: Context<GetPayloadErrorKind>,
+}
+
+impl GetPayloadError {
+    /// Return the kind of this error.
+    pub fn kind(&self) -> &GetPayloadErrorKind {
+        self.ctx.get_context()
+    }
+
+    pub(crate) fn decrypt() -> GetPayloadError {
+        GetPayloadError::from(GetPayloadErrorKind::Decrypt)
+    }
+
+    pub(crate) fn incomplete(needed: Needed, payload: Vec<u8>) -> GetPayloadError {
+        GetPayloadError::from(GetPayloadErrorKind::IncompletePayload { needed, payload })
+    }
+
+    pub(crate) fn deserialize(error: NomErrorKind, payload: Vec<u8>) -> GetPayloadError {
+        GetPayloadError::from(GetPayloadErrorKind::Deserialize { error, payload })
+    }
+}
+
+impl Fail for GetPayloadError {
```

Comment:
Can `derive Fail` help us to reduce this code?

Reply:
This test fails.
https://github.com/tox-rs/tox/blob/8a6cef4d4ec05e0f37f6ddf5d84c1c044f6a87a7/src/toxcore/dht/packet/errors.rs#L118

Don't we need `backtrace()`?

Reply:
I thought this methods is autogenerated with `derive Fail`.